plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java

```diff
@@ -17,60 +17,84 @@
 package com.google.cloud.tools.eclipse.usagetracker;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
-import org.eclipse.swt.widgets.Display;
+import com.google.cloud.tools.eclipse.test.util.ui.CompositeUtil;
+import com.google.cloud.tools.eclipse.test.util.ui.ShellTestResource;
+import org.eclipse.jface.window.Window;
+import org.eclipse.swt.events.FocusAdapter;
+import org.eclipse.swt.events.FocusEvent;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotButton;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
 
 public class OptInDialogTest {
 
-  private final OptInDialog dialog = new OptInDialog(null);
+  @Rule public ShellTestResource shellResource = new ShellTestResource();
+
+  private Shell shell;
+  private OptInDialog dialog;
+
+  @Before
+  public void setUp() {
+    shell = shellResource.getShell();
+    dialog = new OptInDialog(shell);
+  }
 
   @Test
-  public void testIsOptInYes_inUiThread() throws InterruptedException {
-    assertTrue(Display.getCurrent() != null);
-    try {
-      dialog.isOptInYes();
-      fail();
-    } catch (IllegalStateException ex) {
-      assertEquals("Cannot be called from the UI thread.", ex.getMessage());
-    }
+  public void testInitialReturnCode() {
+    assertEquals(Window.CANCEL, dialog.getReturnCode());
   }
 
   @Test
-  public void testIsOptInYes_okPressed() throws InterruptedException {
-    dialog.okPressed();
-    assertTrue(getIsOptInYes());
+  public void testReturnCode_okPressed() {
+    scheduleClosingDialogAfterOpen(CloseAction.PRESS_OK);
+    assertEquals(Window.OK, dialog.open());
   }
 
   @Test
-  public void testIsOptInYes_cancelPressed() throws InterruptedException {
-    dialog.cancelPressed();
-    assertFalse(getIsOptInYes());
+  public void testReturnCode_cancelPressed() {
+    scheduleClosingDialogAfterOpen(CloseAction.PRESS_CANCEL);
+    assertNotEquals(Window.OK, dialog.open());
   }
 
   @Test
-  public void testIsOptInYes_dialogDismissed() throws InterruptedException {
-    dialog.cancelPressed();
-    assertFalse(getIsOptInYes());
+  public void testReturnCode_dialogClosed() {
+    scheduleClosingDialogAfterOpen(CloseAction.CLOSE_SHELL);
+    assertNotEquals(Window.OK, dialog.open());
   }
 
-  private boolean getIsOptInYes() throws InterruptedException {
-    final Boolean[] answerHolder = new Boolean[1];
-    Thread thread = new Thread(new Runnable() {
+  private enum CloseAction { PRESS_OK, PRESS_CANCEL, CLOSE_SHELL };
```

Comment:
Should you test for `dispose()` (`DISPOSE_SHELL`)?  I guess it doesn't matter now with your blocking-dialog pattern.

Reply:
Tried to simulate disposing the shell, but I faced some difficulty with it. Let's settle with `CLOSE_SHELL`.